specs/ChainSpec/Consensus.md

```diff
@@ -0,0 +1,146 @@
+# Consensus
+
+The fields in the Block header relevant to the consensus process are:
+
+```rust
+struct BlockHeader {
+    ...
+    prev_hash: BlockHash,
+    height: BlockHeight,
+    epoch_id: EpochId,
+    last_final_block_hash: BlockHash,
+    approvals: Vec<Approval>
+    ...
+}
+```
+
+Each block belongs to a particular epoch, and has a particular height. The highest height block accepted by a particular node is its head.
+
+Each epoch has a set of block producers who are assigned to produce blocks in the epoch, and has an assigned block producer for each height in the epoch. The block producers and the assignments in a particular epoch are known from the first block of the *preceding* epoch.
+
+Consecutive blocks do not necessarily have sequential heights. A block at height `h+1` can have as its previous block a block with height `h-1` or lower.
+
+If a block `B` at some height `h` builds on top of a block `B_prev` that has height `h-1`, and `B_prev` in turn builds on top of a block `B_prev_prev` that has height `h-2`, and all three blocks are in the same epoch, the block `B_prev_prev` is final, and cannot be reverted unless block producers with more than `2/3` cumulative stake deviate from the protocol. 
+
+TODO: talk about epoch boundaries here
+
+Block producers in the particular epoch exchange many kinds of messages. The two kinds that are relevant to the consensus are **Blocks** and **Approvals**. The approval contains five fields:
+
+```rust
+struct Approval {
+    prev_hash: BlockHash,
+    prev_height: BlockHeight,
+    target_height: BlockHeight,
+    signature: Signature,
+    account_id: AccountId
+}
+```
+
+Where `prev_hash` and `prev_height` are the hash and the height of the block being approved, `target_height` is the specific height at which the approval can be used (an approval with a particular `target_height` can be only included in the `approvals` of a block that has `height = target_height`), `account_id` is the account of the block producer who created the approval, and `signature` is their signature on the tuple `(prev_hash, target_height)` if `target_height == prev_height+1` and the tuple `(prev_height, target_height)` otherwise.
+
+Any valid block must contain approvals from block producers whose cumulative stake exceeds 2/3 of the total stake in the epoch. For a block `B` and its previous block `B'` each approval in `B` must have `prev_height` to be equal to the height of `B'`. If (and only if) `B.height == B'.height + 1`, each approval's `prev_hash` must also be exactly the hash of `B'`. See [this section](#approval-condition) below for details on why the condition must be exactly such.
```

Comment:
Why is this condition different from the one described in Doomslug paper? In the paper, if `B.height == B'.height + 1`, a witness can contain both endorsements and skips, and if `B.height > B'.height + 1`, it can contain skips of different lengths.